### PR TITLE
Add support for subscribing to updates by device MAC

### PR DESCRIPTION
**Pull request by thecode**
Add options to subscribe for updates using device MAC (device ID)
Subscribing using device ID will help to support devices behind NAT, proxies, or if device IP changes
Implementation is backward compatible and allow subscribing for updates either by MAC or device IP, we can later drop the support for subscribing by IP.

Closes https://github.com/home-assistant-libs/aioshelly/issues/182

Gen1:
- CoAP contains the device ID in CoAP option 3332, for this parsing of options was added, logic for parsing options is based on https://github.com/chrysn/aiocoap/blob/master/aiocoap/options.py
- For compatibility with older Gen1 devices that only include the last 6 chars of MAC, comparison only compare the last 6 MAC chars

Gen2:
- Device ID is sent on every frame as `src`

Checked on multiple Gen1/Gen2 devices with various firmware versions for Gen1.
Update `example.py` to test subscribing by MAC, for testing with debug enabled:
```
python example.py --ip 192.168.1.75 -i -deb -m "ABCDEF123456"
```

### CODE REVIEW

File: aioshelly/block_device/coap.py [hunk 1]
```diff
@@ -36,9 +39,34 @@ def __init__(self, sender_addr: tuple[str, int], payload: bytes) -> None:
...
+            self.options[option_number] = raw_data[:length]
+            raw_data = raw_data[length:]
+
+        if not data:
+            raise InvalidMessage("Received message without data")
+
         try:
-            self.payload = json.loads(payload.rsplit(b"\xff", 1)[1].decode())
-        except (json.decoder.JSONDecodeError, UnicodeDecodeError, IndexError) as err:
+            self.payload = json.loads(data.decode())
```
Comment (bdraco) [inline aioshelly/block_device/coap.py:68]:
```suggestion
            self.payload = json_loads(data.decode())
```

Also needs
`from ..json import json_dumps, json_loads`

**Comment by bdraco** [comment]
All my Shelly Plus Plug US work fine with this change

**Review by bieniu — APPROVED** [review]
1L and Plug S work fine, looks good :+1:

**Comment by thecode** [comment]
Thanks @bieniu @bdraco

**Pull request merged**